Files/FileList: adding the same file twice should not increment `FileList::$numFiles`

rodrigoprimo opened the pull request:
# Description

This PR fixes the `FileList::__construct()` and `FileList::addFile()` methods to ensure that, when there is an attempt to add the same file twice, the `FileList::$numFiles` variable is not incremented. This behavior was causing a fatal error when using the `--parallel` option and passing the same file twice (see #1113).

The code was already handling duplicates correctly in the sense that duplicated files were not added to `FileList::$files`. However, `FileList::$numFiles` was being incorrectly incremented.

This PR also includes some basic tests for `FileList::__construct()` and `FileList::addFile()` in separate commits.

## Suggested changelog entry

Fixed: fatal error when using the `--parallel` option and passing the same file twice

## Related issues/external references

Fixes #1113


## Additional notes

There is some duplicated logic in `FileList::__construct()` and `FileList::addFile()`. I considered refactoring the duplicated code to a private method before adding the check that is added in this commit. I decided not to do it as there are some subtle differences between the logic in the two methods that make the refactor not straightforward.

`FileList::__construct()` always sets the value of an entry in the `FileList::$files` to `null` and the key to whatever is returned by `SplFileInfo::getPathname()`. While `FileList::addFile()` sets the value of an entry in the `FileList::$files` to `null` or an object passed to the method, and the key to the path passed to the method.

I decided to leave this refactor to remove the duplication to the future and focus this commit on fixing the issue with handling duplicated files. I believe it will be necessary to write additional tests for this class and also incorporate more defensive code.


## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement


## PR checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] \[Required for new sniffs\] I have added XML documentation for the sniff.

<!--
============================================================================================
Please make sure your pull request passes all continuous integration checks!

PRs which are failing their CI checks will likely be ignored by the maintainers.

Small PRs using atomic, descriptive commits are hugely appreciated as it will make
reviewing your changes easier for the maintainers.
============================================================================================
-->

jrfnl commented on tests/Core/Files/FileList/AddFileTest.php:
```diff
@@ -0,0 +1,137 @@
+            'STDIN'                         => ['fileName' => 'STDIN'],
+            'Regular file with file object' => [
+                'fileName'   => 'test1.php',
+                'fileObject' => new stdClass(),
```
This should create a `File` object, not a `stdClass` object. As things are, this will break in PHPCS 4.0 when the type declarations are added (`addFile(string $path, ?File $file=null)`).

rodrigoprimo replied:
Good catch! I added a commit creating an instance of the `File` class. To do that, I had to instantiate both `ConfigDouble` and `Ruleset`. Let me know if you had a different approach in mind.

rodrigoprimo commented on tests/Core/Files/FileList/AddFileTest.php:
```diff
@@ -0,0 +1,137 @@
+/**
+ * Tests for the \PHP_CodeSniffer\Files\FileList::addFile method.
+ *
+ * @copyright 2019-2025 PHPCSStandards Contributors
```
I copied the line `@copyright 2019-2024 PHPCSStandards Contributors` from another file and, without thinking too much about it, updated it to 2025 since this is the current year.

It is not clear to me if the years apply to just this file or to the whole repository. If it is just this file, should it be?

```
@copyright 2025 PHPCSStandards Contributors
```

I can see that phpDocumentor leaves the format of the description of the `@copyright` tag to each individual project (https://docs.phpdoc.org/guide/references/phpdoc/tags/copyright.html). Maybe this is documented somewhere in this repo, and I missed it?

jrfnl replied:
> It is not clear to me if the years apply to just this file or to the whole repository. If it is just this file, should it be?

Well, it's a bit messy in this repository, so generally speaking, you can assume it applies to the specific file. I still intend to look into streamlining this at some point, but there may be legal implications of doing that, so something I want to be careful with.

`@copyright 2025 PHPCSStandards Contributors` is what I would have expected for this new file.

rodrigoprimo replied:
Done. I opted to make this change in two separate commits (one for ConstructTest.php and FilterDouble.php, and another for AddFileTest.php) to hopefully make it easier when squashing these changes into the original commits.

rodrigoprimo commented on tests/Core/Files/FileList/ConstructTest.php:
```diff
@@ -0,0 +1,129 @@
+
+        $this->assertCount(count($expectedFiles), $fileList, 'File count mismatch');
+
+        $i = 0;
```
My idea is that it would help identify which file caused the assertion to fail by providing the index of the corresponding value in the `$expectedFiles` array. But I can see that this is not very straightforward. Perhaps it would be better to modify the error message to include the `$filePath` instead of `$i`??

jrfnl replied:
I like that idea. Sounds more descriptive to me.

rodrigoprimo commented on tests/Core/Files/FileList/ConstructTest.php:
```diff
@@ -0,0 +1,148 @@
+    /**
+     * Setup objects needed for the tests.
+     *
+     * @before
```
I have a vague memory of us discussing that creating the `Ruleset` class in the tests has a performance impact. That is why I went with the mock of this class. I don't think I'm able to create a mock of `Ruleset` if I use `@beforeClass`.

That being said, I'm not sure what is better: creating an instance of `ConfigDouble` and a mock of `Ruleset` for every single test, or creating an instance of `ConfigDouble` and an instance of `Ruleset` once for all the tests.

Is this more or less what you had in mind with this question?

jrfnl replied:
What I was hinting at was that both the a `beforeClass` method and the data provider need to be static methods - and need to write to static properties, so the `beforeClass` method could check if the properties had been initialized and if not, could create the objects and the data provider could call the `beforeClass` method to do the same and then in both the tests as well as the data provider, the statically stored objects can be re-used.

rodrigoprimo replied:
Thanks for the additional details, Juliette! I have now implemented this change.

rodrigoprimo commented on tests/Core/Files/FileList/AddFileTest.php:
```diff
@@ -0,0 +1,105 @@
+     * @return array<string, array<string>>
+     */
+    public static function dataAddFile()
+    {
```
Done

rodrigoprimo commented:
Thanks for your review, @jrfnl!

> Only real question I have is with regards to the solution applied in the third commit versus the suggestion I left in https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/1113#issuecomment-2941233392
>
>> I've now started wondering if $numFiles shouldn't be set just once at the end of the __construct() method via a count($this->files) instead of doing the incrementing whenever a new file is being added... ? I'd expect that would solve the issue too.
>
> While I can see that, as the addFile() method is public, this wouldn't solve anything for addFile() when not called from the constructor, I do imagine the above suggestion could simplify the code change for the constructor itself.
Was there a particular reason why you went with a different solution ?

I did consider your suggestion, but noticed that it did not apply to `addFile()`. What I failed to realize, is that I could apply it only to the `__construct()` method. I have now added a new commit simplifying how `FileList::$numFiles` is calculated in the constructor.

Besides that, I addressed all the other comments. Could you please take another look when you get a chance?

Also, before merging this PR, let me know if you prefer to squash the commits yourself or if you would like me to do it.

jrfnl commented on tests/Core/Files/FileList/ConstructTest.php:
```diff
@@ -0,0 +1,129 @@
+
+        $fileList = new FileList($this->config, $this->ruleset);
+
+        $this->assertSame($this->config, $fileList->config, 'Config object mismatch');
+        $this->assertSame($this->ruleset, $fileList->ruleset, 'Ruleset object mismatch');
```
I'm fine with leaving the assertions in place, they don't do any harm.

It's more a case of what did you intend to test ? If it is about safeguarding that the properties are being set from the constructor, then checking with an `assertInstanceOf()` could be sufficient. `assertSame()` implies you want to check that the instance is completely the same, prohibiting any changing of those object from within the constructor.

jrfnl commented:
@rodrigoprimo As I'm going to need to look at the complete thing again before merging this, feel free to reorganize the commits already when you make your final changes.

jrfnl commented on tests/Core/Files/FileList/AbstractFileListTestCase.php:
```diff
@@ -0,0 +1,54 @@
+     */
+    public static function initializeConfigAndRuleset()
+    {
+        if (isset(self::$ruleset) === false) {
```
```suggestion
        // Wrapped in an `isset()` as the properties may have been set already (via a call to this method from a dataprovider).
        if (isset(self::$ruleset) === false) {
```

jrfnl reviewed (commented):
Thanks for making those updates @rodrigoprimo!

Just one small suggestion to prevent head scratching in the future. Otherwise this looks good to go.

jrfnl reviewed (approved):
Thanks @rodrigoprimo ! I'll squash that last commit into the first and will merge this once the build has passed.
